File: python/min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.minstack = []
        
    def push(self, x):
        self.stack.append(x)
        if not self.minstack or self.minstack[-1] >= x:
            self.minstack.append(x)
            
    def pop(self):
        if not self.stack:
            raise ValueError('stack underflow')
        val = self.stack.pop()
        if val == self.minstack[-1]:
            self.minstack.pop()
        return val
        
    def top(self):
        if not self.stack:
            raise ValueError('stack underflow')
        return self.stack[-1]
        
    def getMin(self):
        if self.minstack: return self.minstack[-1]

File: java/MinStack.java
import java.util.*;

class MinStack {
    private Stack<Integer> stack = new Stack<Integer>();
    private Stack<Integer> minStack = new Stack<Integer>();

    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty() || x <= minStack.peek()) minStack.push(x);
    }

    // here's a bug of OJ its own, return type should be `int`
    public int pop() {
        int x = stack.pop();
        if (x == minStack.peek()) minStack.pop();
        return x;
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}